### ISSUE TITLE
Basic Auth not working in v.1.7 for https

### ISSUE THREAD

**Issue by dschuldt** [haproxytech/kubernetes-ingress#401]
Hi,

thanks for HAProxy Ingress; an amazing software! Unfortunately with v. 1.7.x (tried 1.7.1 to 1.7.3) our basic auth does not work anymore for https urls.

**K8s Version**: Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.5", .....
**HAProxy Ingress Version**:  haproxytech/kubernetes-ingress:1.7.3

**relevant Ingress Config**:

```
apiVersion: networking.k8s.io/v1
kind: Ingress
...
annotations:
    haproxy.com/auth-type: basic-auth
    haproxy.com/auth-secret: mynamespace/mysecret
...
```
When executing requests, the user is not asked for basic auth credentials.
Having a look at the generated haproxy.cfg, the directive is missing:

> / $ cat /etc/haproxy/haproxy.cfg
```
frontend http 
  mode http
  bind 0.0.0.0:80 name v4
  ....
  http-request auth realm Protected-Content if .....
  ```
The http-request auth line is missing for https frontend. 
Did we miss anything in v.1.7.x documentation?

Thanks in advance

Best

Denis

**Comment by dschuldt** [comment]
@ivanmatmati, thanks for your quick response!

I can confirm it works with plaintext http.
Today I tried with a downgraded haproxy ingress v. 1.6.7. Here basic auth works.

frontends from generated /etc/haproxy/haproxy.cfg **v1.6.7**:

```
frontend http 
  mode http
  bind 0.0.0.0:80 name v4
  http-request set-var(txn.base) base
  http-request set-var(txn.path) path
  http-request set-var(txn.host) req.hdr(Host),field(1,:),lower
  http-request set-var(txn.host_match) var(txn.host),map(/etc/haproxy/maps/host.map)
  http-request set-var(txn.host_match) var(txn.host),regsub(^[^.]*,,),map(/etc/haproxy/maps/host.map,'') if !{ var(txn.host_match) -m found }
  http-request set-var(txn.path_match) var(txn.host_match),concat(,txn.path,),map(/etc/haproxy/maps/path-exact.map)
  http-request set-var(txn.path_match) var(txn.host_match),concat(,txn.path,),map_beg(/etc/haproxy/maps/path-prefix.map) if !{ var(txn.path_match) -m found }
  http-request auth realm "Protected Content" if { var(txn.path_match) -m dom 6a5db987e0485a59de2a925f2123ed7f } !{ http_auth_group(kubernetes-dashboard-kubernetes-dashboard) authenticated-users }
  http-request capture hdr(Host) len 128 if { var(txn.path_match) -m dom 6b75b4bdfd8ffe79dd4959687330636f }
  http-request capture hdr(X-Forwarded-For) len 128 if { var(txn.path_match) -m dom ff02d203eb8de1c2e055f6567daef3ec }
  http-request redirect location https://%[hdr(host),field(1,:)]:443%[capture.req.uri] code 302 if { var(txn.path_match) -m dom 379203a70eb455857c03a27fce61223d }
  use_backend %[var(txn.path_match),field(1,.)]
  default_backend haproxy-controller-ingress-default-backend-http

frontend https 
  mode http
  bind 0.0.0.0:443 name v4 crt /etc/haproxy/certs/frontend ssl alpn h2,http/1.1
  http-request set-var(txn.base) base
  http-request set-var(txn.path) path
  http-request set-var(txn.host) req.hdr(Host),field(1,:),lower
  http-request set-var(txn.host_match) var(txn.host),map(/etc/haproxy/maps/host.map)
  http-request set-var(txn.host_match) var(txn.host),regsub(^[^.]*,,),map(/etc/haproxy/maps/host.map,'') if !{ var(txn.host_match) -m found }
  http-request set-var(txn.path_match) var(txn.host_match),concat(,txn.path,),map(/etc/haproxy/maps/path-exact.map)
  http-request set-var(txn.path_match) var(txn.host_match),concat(,txn.path,),map_beg(/etc/haproxy/maps/path-prefix.map) if !{ var(txn.path_match) -m found }
  http-request auth realm "Protected Content" if { var(txn.path_match) -m dom 6a5db987e0485a59de2a925f2123ed7f } !{ http_auth_group(kubernetes-dashboard-kubernetes-dashboard) authenticated-users }
  http-request capture hdr(Host) len 128 if { var(txn.path_match) -m dom 6b75b4bdfd8ffe79dd4959687330636f }
  http-request capture hdr(X-Forwarded-For) len 128 if { var(txn.path_match) -m dom ff02d203eb8de1c2e055f6567daef3ec }
  http-request set-header X-Forwarded-Proto https
  use_backend %[var(txn.path_match),field(1,.)]
  default_backend haproxy-controller-ingress-default-backend-http
```

here from **v1.7.3**:

```
frontend http
  mode http
  bind 0.0.0.0:80 name v4
  http-request set-var(txn.base) base
  http-request set-var(txn.path) path
  http-request set-var(txn.host) req.hdr(Host),field(1,:),lower
  http-request set-var(txn.host_match) var(txn.host),map(/etc/haproxy/maps/host.map)
  http-request set-var(txn.host_match) var(txn.host),regsub(^[^.]*,,),map(/etc/haproxy/maps/host.map,'') if !{ var(txn.host_match) -m found }
  http-request set-var(txn.path_match) var(txn.host_match),concat(,txn.path,),map(/etc/haproxy/maps/path-exact.map)
  http-request set-var(txn.path_match) var(txn.host_match),concat(,txn.path,),map_beg(/etc/haproxy/maps/path-prefix.map) if !{ var(txn.path_match) -m found }
  http-request auth realm Protected-Content if { var(txn.path_match) -m dom debb735bc99fcbd70255e35de727ed40 } !{ http_auth_group(kubernetes-dashboard-kubernetes-dashboard) authenticated-users }
  http-request capture hdr(Host) len 128
  http-request capture hdr(X-Forwarded-For) len 128
  http-request redirect location https://%[hdr(host),field(1,:)]:443%[capture.req.uri] code 302 if { var(txn.path_match) -m dom 379203a70eb455857c03a27fce61223d } 
  use_backend %[var(txn.path_match),field(1,.)]
  default_backend haproxy-controller-ingress-default-backend-http

frontend https
  mode http
  bind 0.0.0.0:443 name v4 crt /etc/haproxy/certs/frontend ssl alpn h2,http/1.1
  http-request set-var(txn.base) base
  http-request set-var(txn.path) path
  http-request set-var(txn.host) req.hdr(Host),field(1,:),lower
  http-request set-var(txn.host_match) var(txn.host),map(/etc/haproxy/maps/host.map)
  http-request set-var(txn.host_match) var(txn.host),regsub(^[^.]*,,),map(/etc/haproxy/maps/host.map,'') if !{ var(txn.host_match) -m found }
  http-request set-var(txn.path_match) var(txn.host_match),concat(,txn.path,),map(/etc/haproxy/maps/path-exact.map)
  http-request set-var(txn.path_match) var(txn.host_match),concat(,txn.path,),map_beg(/etc/haproxy/maps/path-prefix.map) if !{ var(txn.path_match) -m found }
  http-request capture hdr(Host) len 128
  http-request capture hdr(X-Forwarded-For) len 128
  http-request set-header X-Forwarded-Proto https
  use_backend %[var(txn.path_match),field(1,.)]
  default_backend haproxy-controller-ingress-default-backend-http
```

v1.7.x does not generate the 

> http-request auth...
line.

Since there are no sensitive information I could as well post the ingress manifest, but there is really nothing special in there apart from the haproxy annotations as posted in my initial question.

I'm gonna try it with a minimal nginx deployment right now to see if any other component in our stack messes with the config but I dont think so.

Again, thanks for help.

**Comment by dschuldt** [comment]
@ivanmatmati here's an minimal example, in which basic auth works for http but not https. Ingress version 1.7.3.

```
---
kind: Deployment
apiVersion: apps/v1
metadata:
  name: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      k8s-app: nginx
  template:
    metadata:
      labels:
        k8s-app: nginx
    spec:
      containers:
        - name: nginx
          image: nginx:stable-alpine
          imagePullPolicy: Always
          ports:
            - containerPort: 80
              protocol: TCP
---
apiVersion: v1
kind: Service
metadata:
  name: nginx
spec:
  type: ClusterIP
  selector:
    k8s-app: nginx
  ports:
    - port: 80
      protocol: TCP
      targetPort: 80
---
apiVersion: v1
kind: Secret
type: Opaque
metadata:
  name: haproxy-credentials
data:
  bob: JDEkTFJIM1dEOHAkTzM3RFdEQzVUOEM3QlNhQjFPcnltMA==
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: nginx
  annotations:
    haproxy.com/auth-type: basic-auth
    haproxy.com/auth-secret: haproxy-credentials
spec:
  tls:
    - hosts:
        - my.k8s.fqdn
      secretName: my-tls-secret
  rules:
    - host: my.k8s.fqdn
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: nginx
                port:
                  number: 80

```

> curl http://my.k8s.fqdn

```
<html><body><h1>401 Unauthorized</h1>
You need a valid user and password to access this content.
</body></html>

```

> curl https://my.k8s.fqdn

```
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>
```

The second curl should also expect a 401 Unauthorized.

**Comment by ivanmatmati** [comment]
Thanks, will check it. Check on code didn't pay back.

**Comment by ivanmatmati** [comment]
HI,are you sure that the secret my-tls-secret exists or is reachable ? Indeed, if it doesn't the "http-request auth realm Protected-Content" indeed appears only in http frontend but not in https. That's the sole combination I've seen to reproduce your observation.

**Comment by ivanmatmati** [comment]
Hi, it seems I can now reproduce your observation. The presence of a TLS secret in ingress deactivates the basic auth configuration in HTTPS frontend. Will check if this is intended behavior.

**Comment by ivanmatmati** [comment]
Just to let you know I've seen the source of issue. A fix is implemented but I must wait for confirmation that the behavior you expected is ok with us (very likely to be). You should have a feedback on Wednesday.

**Comment by Mo3m3n** [comment]
Shoud be fixed in `v1.7.4` via https://github.com/haproxytech/kubernetes-ingress/commit/787733194a3284445dbcd0bbbbf5615c56b28a49 
Don't hesitate to re-open this issue if it is not the case.

**Comment by dschuldt** [comment]
@Mo3m3n 
Thanks a lot; I can confirm.
